bff-node/src/config/passport.ts

```diff
@@ -0,0 +1,158 @@
+
+export const initPassport = async () => {
+  if (!SessionRepository) throw new Error('SessionRepository not initialized');
+  if (!ProfileRepository) throw new Error('ProfileRepository not initialized');
+  if (!process.env.ISSUER_URL) {
+    console.error('No issuer url found in environment variables, exiting');
+    process.exit(1);
+  }
+
+  const idportenIssuer = await Issuer.discover(process.env.ISSUER_URL).then(
+    async (idportenIssuer) => {
+      return idportenIssuer;
+    }
+  );
+
+  const client = new idportenIssuer.Client({
+    client_id: process.env.CLIENT_ID!,
+    client_secret: process.env.CLIENT_SECRET,
+    redirect_uris: [process.env.REDIRECT_URI!],
+    response_types: ['code'],
+    token_endpoint_auth_method: 'client_secret_basic',
+    scope: process.env.SCOPE,
+    authorizationURL: process.env.AUTH_URL,
+  });
+
+  passport.use(
+    'oidc',
+    new OpenIDStrategy(
+      { client, passReqToCallback: true },
+      async (req: any, tokenSet: TokenSet, profile: any, done: any) => {
+        try {
+          const idPortenProfile: IDPortenProfile = profile;
+          const postLoginRedirectUrl = req.session.returnTo || '/';
+          const userId = idPortenProfile.pid;
+          const { sid, locale, pid } = tokenSet.claims();
+
+          const user: Profile | null = await getOrCreateProfile(userId, {
+            language: locale || '',
+          });
+          const existingSessionIdFromCookie = readCookie(req);
+
+          if (!user) throw new Error('Fatal: Unable to find/create user');
+          const session = await getOrCreateSession(
+            sid as string,
+            tokenSet,
+            user,
+            existingSessionIdFromCookie,
+            postLoginRedirectUrl
+          );
+          if (!session) throw new Error('Fatal: Unable to find/create session');
+
+          return done(null, {
+            sid,
+            pid,
+            userId,
+            postLoginRedirectUrl,
+            currentSessionId: session?.id,
+          });
+        } catch (error) {
+          console.error("authenticate('oidc') error: ", error);
+          return done(error);
+        }
+      }
+    )
+  );
+
+  passport.serializeUser((user: any, done) => {
+    done(null, user.userId);
+  });
+
+  passport.deserializeUser((userId: string, done) => {
+    // Use the ID to find the user in your database
+    ProfileRepository!.findOneBy({ id: userId as string }).then((p) => done(null, p || false));
+  });
+};
+
+const getOrCreateProfile = async (userId: string, userInfo: Partial<Profile>) => {
+  try {
+    let user: Profile | null = await ProfileRepository!.findOne({ where: { id: userId } });
+
+    if (!user) {
+      const newProfile = new Profile();
+      newProfile.id = userId;
+      newProfile.language = userInfo.language || '';
+      user = await ProfileRepository!.save(newProfile);
+    } else {
+    }
+    if (!user) throw new Error('Fatal: User not found and not able to create');
```

Comment:
not needed because of the check above

Reply:
> not needed because of the check above

Det stemmer vel ikke helt. Først sjekker den om user finnes, og hvis ikke prøver den å lage den. Dersom det skjer en feil og user fortsatt ikke er definert så kaster den error.

Eller hva tenker du? :)

Reply:
aha, ja ser det nå, derp fra min side!